File: wcserver/delphi/examples/TriviaServer/trvsrvr.pas
unit TrvSrvr;

interface

Uses
  WinSock;

function ListenThread(s : TSocket) : integer;

implementation

Uses
  Globals,
  Windows,
  SysUtils,
  SpServer;

function GetTriviaQuestion(index : integer; var s : String) : BOOL;
var
  SetEOF : BOOL;
Begin
  result := FALSE;
  SetEOF := FALSE;
  EnterCriticalSection(DataFileMutex);
  while (TRUE) do begin
    if EoF(DataFile) then begin
      if (not SetEOF) then begin
        SetEOF := TRUE;
        CloseFile(DataFile);
        Assign(DataFile, 'wctrivia.dat'); // hack!!!
        Reset(DataFile);
        if IoResult <> 0 then ;
      end else begin
        result := FALSE;
        break;
      end;
    end;
    ReadLn(DataFile, S);
    if (IoResult = 0) and (Length(S) > 0) and (S[1] <> ';') then begin
      result := TRUE;
      break;
    end;
  end;
  LeaveCriticalSection(DataFileMutex);
End;

// TTriviaServer

const Version : String = '1.00';

type
  TTriviaServer = class(TSimpleProtocolServer)
  public
    constructor Create(s : TSocket);
  protected
    procedure SendWelcome; override;
    function DispatchCommand(s : pchar): BOOL; override;
  protected
    function tsBYE: BOOL;
    function tsUSER(s : pchar): BOOL;
    function tsVER: BOOL;
    function tsNEXT: BOOL;
  end;

constructor TTriviaServer.Create(s : TSocket);
begin
  inherited Create(s);
end;

procedure TTriviaServer.SendWelcome;
begin
  SendString('220-Welcome to the Wildcat! Trivia Service');
  SendString('220-Copyright (c) 1996-2004, Santronics Software, Inc. All rights reserved.');
  SendString('220-This is an example of a Wildcat! Trivia Service. This service');
  SendString('220-demonstrates how a Wildcat! Service can be used to provide');
  SendString('220-specific information to callers using a dialup or telnet terminal');
  SendString('220 connection, Dynamic HTML, or the Wildcat! Navigator.');
end;

function TTriviaServer.DispatchCommand(s : pchar): BOOL;
begin
  if (StrLIComp(s, 'BYE', 3) = 0) then begin
    DispatchCommand := tsBYE; exit;
  end else
  if (StrLIComp(s, 'NEXT', 4) = 0) then begin
    DispatchCommand := tsNEXT; exit;
  end else
  if (StrLIComp(s, 'USER', 4) = 0) then begin
    DispatchCommand := tsUSER(nil); exit;
  end else
  if (StrLIComp(s, 'VER', 3) = 0) then begin
    DispatchCommand := tsVER; exit;
  end;
  DispatchCommand := FALSE;
end;

function TTriviaServer.tsBYE: BOOL;
begin
  SendString('221 Goodbye');
  closesocket(socket);
  tsBYE := TRUE;
end;

function TTriviaServer.tsNEXT: BOOL;
var
  S : String;
begin
  if (GetTriviaQuestion(0, s)) then begin
    SendString('200 '+S);
  end else begin
    SendString('500 Error');
  end;
  tsNEXT := TRUE;
end;

function TTriviaServer.tsUSER(s : pchar): BOOL;
begin
  SendString('230 User logged in successfully.');
  tsUSER := TRUE;
end;

function TTriviaServer.tsVER: BOOL;
begin
  SendString('200 wcTriva Version '+Version);
  tsVER := TRUE;
end;

// TriviaThread

function TriviaThread(s : TSocket) : integer;
var
  Server : TTriviaServer;
begin
  //MessageBox(0, 'Starting Thread', '', mb_OK);
  InterlockedIncrement(integer(ActiveConnections));
  Server := TTriviaServer.Create(s);
  Server.Execute;
  Server.Destroy;
  closesocket(s);
  InterlockedDecrement(integer(ActiveConnections));
  result := 0;
end;

// ListenThread

function ListenThread(s : TSocket) : integer;
var
  sin          : TSockAddrIn;
  sinlen       : integer;
  acceptSocket : TSocket;
  THID         : DWORD;
begin
  listen(s, 5);
  while (TRUE) do begin
    sinlen := sizeof(sin);
    acceptSocket := accept(s,
                           @sin,
                           @sinlen);
    BeginThread(nil, 0, @TriviaThread, pointer(acceptSocket), 0, THID);
  end;
end;

end.

File: wcserver/cpp/examples/TriviaServer/trvsrvr.cpp
//***********************************************************************
// (c) Copyright 1998-2008 Santronics Software, Inc. All Rights Reserved.
//***********************************************************************
//
// File Name : trvsrvr.cpp
// Subsystem :
// Date      :
// Version   :
// Author    :
//
// Revision History:
// Build    Date      Author  Comments
// -----    --------  ------  -------------------------------------------
//***********************************************************************

#include "common.h"
#pragma hdrstop

#include "spserver.h"

///////////////////////////////////////////////////////////////////////////////
// GetTriviaQuestion

BOOL GetTriviaQuestion(char *buf, DWORD &size)
{
  BOOL result = FALSE;
  EnterCriticalSection(&DataFileMutex);
  if (DataFile.IsOk()) {
    BOOL seteof = FALSE;
    while (TRUE) {
      DWORD read = size;
      if (DataFile.ReadLine(buf, read)) {
        if ((buf[0] != '\0') && (buf[0] != ';')) {
          size = read;
          result = TRUE;
          break;
        }
      } else
      if ((GetLastError() == ERROR_HANDLE_EOF) && (seteof == FALSE)) {
        seteof = TRUE;
        DataFile.SetFilePointer(0, FILE_BEGIN);
      } else {
        result = FALSE;
        break;
      }
    }
  }
  LeaveCriticalSection(&DataFileMutex);
  return result;
}

///////////////////////////////////////////////////////////////////////////////
// TTriviaServer

class TTriviaServer: public TSimpleProtocolServer {
typedef TSimpleProtocolServer inherited;
public:
  TTriviaServer(SOCKET s);
  ~TTriviaServer();
protected:
  virtual void Go();
  virtual void SendWelcome();
private:
  static TSPDispatch Dispatch[];
  BOOL tsHELP(char *args);
  BOOL tsBYE(char *args);
  BOOL tsUSER(char *args);
  BOOL tsVER(char *args);
  BOOL tsNEXT(char *args);
};


TSimpleProtocolServer::TSPDispatch TTriviaServer::Dispatch[] = {
	SPCMD(TTriviaServer, "HELP",  tsHELP),
	SPCMD(TTriviaServer, "BYE",   tsBYE),
	SPCMD(TTriviaServer, "USER",  tsUSER),
	SPCMD(TTriviaServer, "VER",   tsVER),
	SPCMD(TTriviaServer, "NEXT",  tsNEXT),
	{0}
};

TTriviaServer::TTriviaServer(SOCKET s)
 : TSimpleProtocolServer(s, Dispatch)
{
  Done = FALSE;
  InterlockedIncrement((long*)&ActiveConnections);
  Start();
}

TTriviaServer::~TTriviaServer()
{
  InterlockedDecrement((long*)&ActiveConnections);
}

void TTriviaServer::Go()
{
  inherited::Go();
  delete this;
}

void TTriviaServer::SendWelcome()
{
  Send("220-Welcome to the Wildcat! Trivia Service\r\n");
  Send("220-Copyright (c) 1996-2023, Santronics Software, Inc. All rights reserved.\r\n");
  Send("220-This is an example of a Wildcat! Trivia Service. This service\r\n");
  Send("220-demonstrates how a Wildcat! Service can be used to provide\r\n");
  Send("220-specific information to callers using a dialup or telnet terminal\r\n");
  Send("220-connection, Dynamic HTML, or the Wildcat! Navigator.  Client applications\r\n");
  Send("220 can be written in any language supported by the WCSDK!\r\n");
}

BOOL TTriviaServer::tsBYE(char *args)
{
  Send("221 Goodbye\r\n");
  closesocket(Control);
  Done = TRUE;
  return TRUE;
}

BOOL TTriviaServer::tsNEXT(char *args)
{
  char buf[256];
  DWORD size = sizeof(buf);
  if (GetTriviaQuestion(buf, size)) {
    Send("200 %s\r\n", buf);
  } else {
    Send("500 Error\r\n");
  }
  return TRUE;
}

BOOL TTriviaServer::tsUSER(char *args)
{
  Send("230 User logged in successfully.\r\n");
  return TRUE;
}

BOOL TTriviaServer::tsVER(char *args)
{
  Send("200 wcTrivia Version %s\r\n", WCTRIVIA_VERSION_STR);
  return TRUE;
}

BOOL TTriviaServer::tsHELP(char *args)
{
  Send("200------  wcTrivia Available commands ------------\r\n");
  Send("200-HELP   - gets the next trivia question\r\n");
  Send("200-NEXT   - gets the next trivia question\r\n");
  Send("200-USER   - pass the user name, info\r\n");
  Send("200-VER    - shows wcTrivia version\r\n");
  Send("200-BYE    - exits current session\r\n");
  Send("200 End of Commands\r\n");
  return TRUE;
}

////////////////////////////////////////////////////////////
// ListenThread()
//
// This is the main service TCP listening thrad
////////////////////////////////////////////////////////////

void ListenThread(void *p)
{
  SOCKET s = (SOCKET)p;
  listen(s, 5);
  while (1) {
    sockaddr_in tin;
    int n = sizeof(tin);
    SOCKET t = accept(s, (sockaddr *)&tin, &n);
    if (t == INVALID_SOCKET) {
      break;
    }
    new TTriviaServer(t);
  }
}
